### PR TITLE
feat(eslint-plugin): add new rule require-await

### LINKED ISSUE #669: Rule proposal: typescript-eslint version of the base [require-await] rule
**Repro**

Given the following config, it seems that there is an unresolvable conflict:

```JSON
{
  "rules": {
    "require-await": "error",
    "no-return-await": "error",
    "@typescript-eslint/promise-function-async": "error"
  }
}
```

A function that returns a promise, such as:

```TS
function numberOne(): Promise<number> {
  return Promise.resolve(1);
}
```

..triggers `@typescript-eslint/promise-function-async` with the warning:

```
Functions that return promises must be async
```

Adding the `async` keyword:

```TS
async function numberOne(): Promise<number> {
  return Promise.resolve(1);
}
```

..then triggers `require-await` with the warning:

```
async function ’numberOne’ has no ‘await’ expression
```

Finally, adding the `await` keyword:

```TS
async function numberOne(): Promise<number> {
  return await Promise.resolve(1);
}
```

..then triggers `no-return-await` with the warning:

```
Redundant use of ‘await’ on a return value
```

**Expected Result**

No errors

**Actual Result**

(see repro above)

**Additional Info**

Individually, each of the configured rules address a legitimate problem. But together they appear to be mutually exclusive.

Over in ESlint core, there was a discussion on whether an `async` function that returns the result of _another_ `async` function should be allowed to omit the `await` keyword without triggering `require-await`.

As @kaicataldo [points out](https://github.com/eslint/eslint/issues/10829#issuecomment-429945262) though, it isn’t really possible because ESLint is constrained by the limits of static analysis, and suggests that the help of a typing system (e.g Typescript) would be needed to implement this exception.

Hence I’m raising the issue here, to gather feedback on whether this implies the need for a `@typescript-eslint/require-await` version of the base rule.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.10.2` |
| `@typescript-eslint/parser`        | `1.10.2` |
| `TypeScript`                       | `3.5.1` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `12.1.0` |
| `npm`                              | `6.9.0` |

### PR DESCRIPTION
**Pull request by scottohara**
Extends the `require-await` rule from ESlint core to allow for async functions that _explicitly_ return a promise (as opposed to a non-promise value that the `async` function _implicitly_ wraps in `Promise.resolve`).

To recap, the main intent of the `require-await` rule in ESlint core is guard against marking a function as `async` unnecessarily. The presence of an `await` expression is the signal used to determine if `async` is necessary.

However, when used in conjunction with:
* `@typescript-eslint/promise-function-async`, which requires *all* functions that return a promise to be marked as `async`; and
* `no-return-await`, which warns _against_ using `return await`

..there is an unresolvable conflict.

Fixes #669 

### Examples of newly allowed code

```TS
/*eslint require-await: "off"*/
/*eslint @typescript-eslint/require-await: "error"*/

async function numberOne(): Promise<number> {
  return Promise.resolve(1);
}
```

The above function would normally generate a warning from `require-await`:

```
Async function ‘numberOne’ has no `await` expression.
```

Similarly:

```TS
async function numberOne(): Promise<number> {
  return getAsyncNumber(1);
}

async function getAsyncNumber(x: number): Promise<number> {
  return Promise.resolve(x);
}
```

The above `async` function `numberOne` returns the result of _another_ `async` function `getAsyncNumber`, which would normally trigger two warnings.

Aside from this one specific case, the rule passes through to ESLint core `require-await` for everything else, so the following still applies:

### Examples of valid code

```TS
// Non-async function declaration
function numberOne(): number {
  return 1;
}

// Non-async function expression
const numberOne = function(): number {
  return 1;
}

// Non-async arrow function expression
const numberOne = (): number => 1;

// Async function declaration with await
async function numberOne(): Promise<number> {
  return await 1;
}

// Async function expression with await
const numberOne = async function(): Promise<number> {
  return await 1;
}

// Async arrow function expression with await
const numberOne = async (): Promise<number> => await 1;
```

### Examples of invalid code

```TS
// Async function declaration with no await
async function numberOne(): Promise<number> {
  return 1;
}

// Async function expression with no await
const numberOne = async function(): Promise<number> {
  return 1;
}

// Async arrow function expression with no await
const numberOne = async (): Promise<number> => 1;
```

### CODE REVIEW

File: packages/eslint-plugin/tests/rules/require-await.test.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        return Promise.resolve(1);
+      }`,
+    },
+    {
+      // Async function expression with promise return
+      code: `const numberOne = async function(): Promise<number> {
+        return Promise.resolve(1);
+      }`,
+    },
+    /*{
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/rules/require-await.test.ts]: commented out test?

File: packages/eslint-plugin/tests/rules/require-await.test.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    {
+      // Async function expression with async function return
+      code: `const numberOne = async function(): Promise<number> {
+        return getAsyncNumber(1);
+      }
+      const getAsyncNumber = async function(x: number): Promise<number> {
+        return Promise.resolve(x);
+      }`,
+    },
+    /*{
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/rules/require-await.test.ts]: ditto?

File: packages/eslint-plugin/tests/rules/require-await.test.ts (hunk 1 above)
Comment (scottohara) [reply packages/eslint-plugin/tests/rules/require-await.test.ts]: Removed

File: packages/eslint-plugin/tests/rules/require-await.test.ts (hunk 2 above)
Comment (scottohara) [reply packages/eslint-plugin/tests/rules/require-await.test.ts]: Removed

File: packages/eslint-plugin/src/rules/require-await.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+      'ArrowFunctionExpression[async = true]:exit': exitFunction,
+
+      ReturnStatement(node: TSESTree.ReturnStatement) {
+        if (!scopeInfo) {
+          return;
+        }
+
+        const { expression } = parserServices.esTreeNodeToTSNodeMap.get(
+          node,
+        ) as ts.ReturnStatement;
```
[the remark this answers is not in the document]
Comment (scottohara) [reply packages/eslint-plugin/src/rules/require-await.ts]: Done

**Comment by bradzacher** [comment]
unless github is glitching out on me, it looks like your push didn't push any changes.

**Comment by scottohara** [comment]
Sorry @bradzacher, minor brain explosion here.  Should be OK now.

**Comment by fregante** [comment]
I think this rule missed await-less arrow functions:

# OK

```ts
async function foo() {
	return Promise.resolve();
}
```

# Not OK

```
const foo = () => Promise.resolve();
```

**Comment by scottohara** [comment]
@fregant 

The arrow function use case should be addressed in #826. (merged, not yet released)

**Pull request merged**